### ISSUE TITLE
Highlights not returned when checking for the non-existence of a field when it is in the mapping

### ISSUE THREAD

**Issue by pauliclark** [elastic/elasticsearch#93382]
### Elasticsearch Version

8.6.0

### Java Version

_bundled_

### OS Version

Linux (from official Docker Image)

### Problem Description

When fetching the highlight of a query, but query is checking for a field to not exist, the highlight is not returned.
If the field to be checked is not in the mapping, then it is returned.

### Query where **securityDescriptors** is in the mapping

```
POST /my_wonderful_index/_search
{
  "query": {
    "bool": {
      "must": [
        {
          "query_string": {
            "default_operator": "or",
            "fields": [
              "textContent"
            ],
            "lenient": true,
            "query": "buckley",
            "type": "cross_fields"
          }
        },
        {
          "ids": {
            "values": [
              "264a61aeb4487a79ccb54a3071fc2a808ae2911eb2c796d172ce875bf170464a"
            ]
          }
        }
      ],
      "must_not": [
        {
          "exists": {
            "field": "securityDescriptors"
          }
        }
      ]
    }
  },
  "highlight": {
    "encoder": "html",
    "fields": {
      "textContent": {
        "fragment_size": 200
      },
      "textContent.exact": {
        "fragment_size": 200
      }
    },
    "number_of_fragments": 5,
    "post_tags": [
      "</b>"
    ],
    "pre_tags": [
      "<b class=\"highlight\">"
    ]
  },
  "_source": {
    "excludes": [
      "*"
    ]
  }
}
```

Responds with:

```
{
  "took": 2,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 9.82759,
    "hits": [
      {
        "_index": "my_wonderful_index",
        "_id": "264a61aeb4487a79ccb54a3071fc2a808ae2911eb2c796d172ce875bf170464a",
        "_score": 9.82759,
        "_source": {}
      }
    ]
  }
}
```

### Query without the **securityDescriptors** existence check

```
POST /my_wonderful_index/_search
{
  "query": {
    "bool": {
      "must": [
        {
          "query_string": {
            "default_operator": "or",
            "fields": [
              "textContent"
            ],
            "lenient": true,
            "query": "buckley",
            "type": "cross_fields"
          }
        },
        {
          "ids": {
            "values": [
              "264a61aeb4487a79ccb54a3071fc2a808ae2911eb2c796d172ce875bf170464a"
            ]
          }
        }
      ]
    }
  },
  "highlight": {
    "encoder": "html",
    "fields": {
      "textContent": {
        "fragment_size": 200
      },
      "textContent.exact": {
        "fragment_size": 200
      }
    },
    "number_of_fragments": 5,
    "post_tags": [
      "</b>"
    ],
    "pre_tags": [
      "<b class=\"highlight\">"
    ]
  },
  "_source": {
    "excludes": [
      "*"
    ]
  }
}
```

Responds with:

```
{
  "took": 10,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 10.355715,
    "hits": [
      {
        "_index": "my_wonderful_index",
        "_id": "264a61aeb4487a79ccb54a3071fc2a808ae2911eb2c796d172ce875bf170464a",
        "_score": 10.355715,
        "_source": {},
        "highlight": {
          "textContent": [
            """Jason <b class="highlight">Buckley</b>
86 Some Road,
Some City
Some Postal code

Date 02&#x2F;07&#x2F;2019

Dear HR,

Re: Subject matter

I am writing to seek your help in resolving""",
            """Jason <b class="highlight">Buckley</b>, 

Customer Payment Lead, 

IT"""
          ]
        }
      }
    ]
  }
}
```

Which includes the highlighting as expected.

### Query with a **notAMappedField** existence check

Interestingly, if the existence check, checks for a field that is not in the mapping, the highlight is still returned.

```
POST /my_wonderful_index/_search
{
  "query": {
    "bool": {
      "must": [
        {
          "query_string": {
            "default_operator": "or",
            "fields": [
              "textContent"
            ],
            "lenient": true,
            "query": "buckley",
            "type": "cross_fields"
          }
        },
        {
          "ids": {
            "values": [
              "264a61aeb4487a79ccb54a3071fc2a808ae2911eb2c796d172ce875bf170464a"
            ]
          }
        }
      ],
      "must_not": [
        {
          "exists": {
            "field": "notAMappedField"
          }
        }
      ]
    }
  },
  "highlight": {
    "encoder": "html",
    "fields": {
      "textContent": {
        "fragment_size": 200
      },
      "textContent.exact": {
        "fragment_size": 200
      }
    },
    "number_of_fragments": 5,
    "post_tags": [
      "</b>"
    ],
    "pre_tags": [
      "<b class=\"highlight\">"
    ]
  },
  "_source": {
    "excludes": [
      "*"
    ]
  }
}
```

Responds with:

```
{
  "took": 10,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 10.355715,
    "hits": [
      {
        "_index": "my_wonderful_index",
        "_id": "264a61aeb4487a79ccb54a3071fc2a808ae2911eb2c796d172ce875bf170464a",
        "_score": 10.355715,
        "_source": {},
        "highlight": {
          "textContent": [
            """Jason <b class="highlight">Buckley</b>
86 Some Road,
Some City
Some Postal code

Date 02&#x2F;07&#x2F;2019

Dear HR,

Re: Subject matter

I am writing to seek your help in resolving""",
            """Jason <b class="highlight">Buckley</b>, 

Customer Payment Lead, 

IT"""
          ]
        }
      }
    ]
  }
}
```

**Comment by ardeshirj** [comment]
We noticed similar behavior while trying to upgrade ES docker from `8.1.3` to `8.3.2` in local environment. Currently in `8.1.3` having `must_not` field has no impact on highlight result, however in `8.3.2` version the highlights no longer return.  An interesting observation was in live instance version `8.3.2` that run in cluster mode the highlights work just fine!

Similar issue was also reported here https://discuss.elastic.co/t/highlighting-not-working-after-upgrade/312095. Suggested workaround was to set highlight type to `plain`, which helps to get the highlights working again but as we know this type has its own limitations.

**Comment by ardeshirj** [comment]
It is not clear in which versions this issue was resolved, but it is no longer an issue in `8.9.1` version.

**Comment by romseygeek** [comment]
This was a lucene bug which was solved in https://github.com/apache/lucene/pull/12088 and released with elasticsearch 8.7
